cpp/tests/streaming/test_shuffler.cpp

```diff
@@ -200,43 +241,54 @@ Node shuffler_nb(
 
     // extract task: extract the packed chunks from the shuffler and send them to the
     // output channel
-    auto extract_task = [](auto shuffler_ctx, auto ctx, auto ch_out) -> Node {
-        ShutdownAtExit c{
-            ch_out
-        };  // TODO: could this be problematic with multiple consumers?
+    auto extract_task =
+        [](auto shuffler_ctx, auto ctx, auto ch_out, auto& latch) -> Node {
         co_await ctx->executor()->schedule();
 
-        while (!shuffler_ctx->shuffler->finished() || !shuffler_ctx->ready_pids.empty()) {
-            auto expected = co_await shuffler_ctx->ready_pids.pop();
-            if (!expected.has_value()) {  // queue is shutdown, so exit the loop
-                break;
+        while (!shuffler_ctx->shuffler->finished()) {
+            auto pid = co_await shuffler_ctx->ready_pids.pop();
+            if (!pid) {
+                break;  // queue is shutdown, so exit the loop
             }
 
-            auto packed_chunks = shuffler_ctx->shuffler->extract(*expected);
+            auto packed_chunks = shuffler_ctx->shuffler->extract(*pid);
 
             co_await ch_out->send(
-                std::make_unique<PartitionVectorChunk>(
-                    *expected, std::move(packed_chunks)
-                )
+                std::make_unique<PartitionVectorChunk>(*pid, std::move(packed_chunks))
             );
 
             if (shuffler_ctx->shuffler->finished()) {
                 // if the shuffler is finished, shutdown & drain the ready_pids queue
                 co_await shuffler_ctx->ready_pids.shutdown_drain(ctx->executor());
             }
         }
+
+        latch.count_down();  // this task is finished, so count down the latch
+    };
+
+    // shutdown task: shutdown the shuffler after all extract tasks have finished
+    auto shutdown_task =
+        [](auto shuffler_ctx, auto ctx, auto ch_out, auto& latch) -> Node {
+        ShutdownAtExit c{ch_out};
+        co_await ctx->executor()->schedule();
+
+        co_await latch;  // wait for all extract tasks to finish before clean up
         co_await ch_out->drain(ctx->executor());
+
+        shuffler_ctx->shuffler->shutdown();
```

Comment:
This doesn't block because at this point we've extracted everything? I suppose it might block if we are still sending data to other people?